### ISSUE TITLE
github auth does not work with JSON

### ISSUE THREAD

**Issue by joshmarinacci** [stormpath/express-stormpath#420]
I'm trying to implement Github auth using the JSON post as described at the bottom of this page:

https://docs.stormpath.com/nodejs/express/latest/social_login.html#json-api

The docs say I should pass in either accessToken or code, depending on which I receive from the provider.  Github gives me 'code', but posting that to the /login route gives me the error:

error {
    "status": 400,
    "message": "providerData accessToken cannot be null, empty, or blank."
}

Does this mean /login with JSON does not actually support codes, not access tokens?

**Comment by robertjd** [comment]
Hi @jakubswiatczak , my understanding with Github is that they only support the page-based direct flow, so I believe this library is only accepting the access code when the user is redirected back to `/callbacks//github` with a code in the URL.  Is there another mechanism by which you are obtaining the code?  Thanks!

**Comment by joshmarinacci** [comment]
Github redirects back to the url I provided, which in this case is http://localhost:18080/cbs along with a query string. This string contains: 

{code: "88427c6ff494c77fed07", state: "bb3b638d-cb4c-f0ce-b01a-3cd860d96310"}

The /cbs route is in my React UI. When the user reaches this page I'm posting back to my API server to the /login route provided by express-stormpath. That's when I get the errror from the stormpath middleware.

**Comment by robertjd** [comment]
Thanks for the update, we'll have to look into how we handle the JSON post to the /login route.  In the meantime, is there a reason why you can't use the callback handler that this library provides by default?  That information will help us design a better solution.  Thanks!

**Comment by joshmarinacci** [comment]
Are you saying I should have github redirect back to the express server direct, rather than to my react app?  I think that might work, but then the user is leaving the react app and would have to come back again later.

**Comment by joshmarinacci** [comment]
I'm digging through the express-stormpath codebase right now. where is the /login route actually implemented?

**Comment by robertjd** [comment]
Yes, in it's current form it's expected that the user will be redirected through /callbacks/github.  Doesn't the user have to leave your React app, to go to the Github site to provide authorization?

The login route handler can be found in lib/controllers/login.js

**Comment by joshmarinacci** [comment]
True, the user does. I'm still not sure what is really the best way to do Oauth when you have a single page webapp.  Do most people only redir to their API server, then back to the client webapp?

**Comment by robertjd** [comment]
It depends on what OAuth flows are supported by the provider.  Github only supports the authorization code flow (which requires the user to be redirected through an endpoint on your server), whereas Google and Facebook provide the "implicit" flow, which allows you to do everything in your SPA with the JavaScript libraries that they provide.

**Comment by edjiang** [comment]
I'm not an expert in what you can do with JS in the browser, but can't you do some sort of pop-up based flow? Eg, open up a pop up to the GitHub authorization page, it redirects to `/callbacks/linkedin`, which then calls `window.opener.location.href = spaUri + '#access_token=' + access_token` and closes itself? So basically hacking the auth code flow into the implicit flow.

@robertjd it would basically be similar to the flow for mobile we were talking about earlier today.

**Comment by joshmarinacci** [comment]
Okay. I've changed my code and my github settings to that the user is taken to http://localhost:30065/callbacks/github,  which is on my express server.  Now the server renders it's own login dialog with the error message: Token is Invalid.

I've tried deleting users from stormpath and revoking all the github user tokens, same error. I'm unsure how to proceed.
